src/sfptpd_clock.c

```diff
@@ -1095,13 +1117,72 @@ static void clock_record_step(void)
 	clock_lock();
 	for (clock = sfptpd_clock_list_head; clock != NULL; clock = clock->next) {
 		assert(clock->magic == SFPTPD_CLOCK_MAGIC);
-		if (clock->type != SFPTPD_CLOCK_TYPE_SYSTEM)
+		if (!clock->deleted &&
+		    clock->type != SFPTPD_CLOCK_TYPE_SYSTEM &&
+		    clock->u.nic.phc)
 			sfptpd_phc_record_step(clock->u.nic.phc);
 	}
 	clock_unlock();
 }
 
 
+static int ptr_compar(const void *a, const void *b)
+{
+	const void **pa = (const void **) a;
+	const void **pb = (const void **) b;
+
+	if (*pa < *pb)
+		return -1;
+	else if (*pa > *pb)
+		return 1;
+	else
+		return 0;
+}
+
+
+static struct sfptpd_clock **clock_snapshot(size_t *num_clocks, bool no_sfc)
+{
+	struct sfptpd_clock **snapshot;
+	struct sfptpd_clock *node;
+	size_t count;
+	int index;
+
+	count = 0;
+	for (node = sfptpd_clock_list_head; node != NULL; node = node->next) {
+		assert(node->magic == SFPTPD_CLOCK_MAGIC);
+		if (!node->deleted &&
+		    (!no_sfc || node->type == SFPTPD_CLOCK_TYPE_NON_SFC)) count++;
+	}
+
+	snapshot = calloc(count, sizeof *snapshot);
+	if (!snapshot) {
+		ERROR("clock: error allocating space for active clock snapshot, %s\n",
+		      strerror(errno));
+		goto finish;
+	}
+
+	index = 0;
+	for (node = sfptpd_clock_list_head; node != NULL; node = node->next) {
+		assert(node->magic == SFPTPD_CLOCK_MAGIC);
+		if (!node->deleted &&
+		    (!no_sfc || node->type == SFPTPD_CLOCK_TYPE_NON_SFC)) {
+			assert(index < count);
+			snapshot[index++] = node;
+		}
+	}
+	assert(index == count);
```

Comment:
It looks like you're checking the clocks didn't change under us - is it OK to only assert that, since customers won't use debug builds?  Should we explicitly abort instead?  Or at least report that automatic deduplication is failing?  Though it looks like maybe the code always had this possibility.

Reply:
No, this is not about checking for clocks changing underneath us, this assertion is a consistency check as there are two ways of counting here: index into the table we are populating and links through the list. The lock is already supposed to be held - I've added a comment before the function now indicating that this is part of the contract.